pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,306 @@
+            ),
+        )
+        ```
+    """
+
+    tool_descriptions: dict[str, str] = field(default_factory=lambda: {})
+    """Custom descriptions for tools, keyed by tool name."""
+
+    tool_args_descriptions: dict[str, dict[str, str]] = field(default_factory=lambda: {})
+    """Custom descriptions for tool arguments: `{'tool_name': {'arg_name': 'description'}}`."""
+
+    def get_tool_args_for_tool(self, tool_name: str) -> dict[str, str] | None:
+        """Get the tool argument descriptions for the given tool name."""
+        return self.tool_args_descriptions.get(tool_name)
+
+
+@dataclass
+class PromptConfig:
+    """Configuration for customizing all strings and prompts sent to the model by Pydantic AI.
+
+    `PromptConfig` provides a clean, extensible interface for overriding any text that
+    Pydantic AI sends to the model. This includes:
+
+    - **Prompt Templates**: Messages for retry prompts, tool return confirmations,
+      validation errors, and other system-generated text via [`PromptTemplates`][pydantic_ai.PromptTemplates].
+    - **Tool Configuration**: Tool descriptions, parameter descriptions, and other
+      tool metadata - allowing you to override descriptions and args for tools at the agent level.
+
+    This allows you to fully customize how your agent communicates with the model
+    without modifying the underlying tool or agent code.
+
+    Note:
+        At least one of `templates` or `tool_config` must be provided. Creating a
+        `PromptConfig()` with no arguments will raise a `ValueError`.
+
+    Example:
+        ```python
+        from pydantic_ai import Agent, PromptConfig, PromptTemplates
+
+        agent = Agent(
+            'openai:gpt-4o',
+            prompt_config=PromptConfig(
+                templates=PromptTemplates(
+                    validation_errors_retry='Please correct the errors and try again.',
+                    final_result_processed='Result received successfully.',
+                ),
+            ),
+        )
+        ```
+
+    Attributes:
+        templates: Templates for customizing system-generated messages like retry prompts,
+            tool return confirmations, and validation error messages.
+        tool_config: Configuration for customizing tool descriptions and metadata.
+    """
+
+    templates: PromptTemplates | None = None
+    """Templates for customizing system-generated messages sent to the model.
+
+    See [`PromptTemplates`][pydantic_ai.PromptTemplates] for available template options.
+    """
+
+    tool_config: ToolConfig | None = None
+    """Configuration for customizing tool descriptions and metadata.
+    See [`ToolConfig`][pydantic_ai.ToolConfig] for available configuration options.
+    """
+
+    def __post_init__(self):  # pragma: no cover
+        if self.templates is None and self.tool_config is None:
+            raise ValueError(
+                "PromptConfig requires at least 'templates' or 'tool_config' to be provided. "
```

Comment:
Do we really need to require this?

Reply:
I was thinking users initializing an empty PromptConfig() should not have any bias that it will do something and needs at least one of the inner fields to operate something with.

Reply:
Removed after all, wasn't necessary.